File: scripts/importPsspKontrak.py
"""
Import PSSP Pelunasan Excel into PsspKontrak table.

Usage:
    python scripts/importPsspKontrak.py "excel/20260630_PSSP Pelunasan_Biaya Murni dan Pelunasan_Value.xlsx"

Requires: pip install pandas openpyxl psycopg2-binary python-dotenv
"""
import sys, re, json, os, pathlib
from datetime import date, datetime

# ── Load .env ─────────────────────────────────────────────────────────────────
env_path = pathlib.Path(__file__).parent.parent / ".env"
if env_path.exists():
    for line in env_path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if line and not line.startswith("#") and "=" in line:
            k, v = line.split("=", 1)
            os.environ.setdefault(k.strip(), v.strip().strip('"').strip("'"))

import psycopg2
import psycopg2.extras
import pandas as pd

DATABASE_URL = os.environ.get("DATABASE_URL", "")
if not DATABASE_URL:
    sys.exit("DATABASE_URL not set in .env")

# Parse postgres URL: postgresql://user:pass@host:port/db
# Also handle ?... query params
import re as _re
_m = _re.match(r"postgresql(?:\+[^:]+)?://([^:]+):([^@]+)@([^:/]+):?(\d*)/([^?]+)", DATABASE_URL)
if not _m:
    sys.exit(f"Cannot parse DATABASE_URL: {DATABASE_URL}")
_user, _pass, _host, _port, _db = _m.groups()
_port = int(_port) if _port else 5432

EXCEL_PATH = sys.argv[1] if len(sys.argv) > 1 else "excel/20260630_PSSP Pelunasan_Biaya Murni dan Pelunasan_Value.xlsx"

# Extract snapshot date from filename (YYYYMMDD prefix)
_date_match = re.search(r"(\d{8})", pathlib.Path(EXCEL_PATH).name)
SNAPSHOT_DATE = date(int(_date_match.group()[:4]), int(_date_match.group()[4:6]), int(_date_match.group()[6:8])) if _date_match else None

print(f"Reading {EXCEL_PATH} (snapshot {SNAPSHOT_DATE})...")
df = pd.read_excel(
    EXCEL_PATH, sheet_name="PSSP",
    dtype={"KD_PRODUK": str, "KD_OUTLET": str, "KD_CUST": str, "C_URUT": str, "KD_SPC": str},
)
print(f"  {len(df)} rows loaded")
before = len(df)
df = df[df["C_URUT"].notna() & (df["C_URUT"].astype(str).str.strip() != "")]
print(f"  {len(df)} rows with valid C_URUT (skipped {before - len(df)} null)")

# ── Column helpers ──────────────────────────────────────────────────────────────
def period_cols(prefix):
    return sorted([c for c in df.columns if re.match(rf"^{prefix}_\d{{6}}$", c)])

EST_COLS  = period_cols("Est")
BM_COLS   = period_cols("BM")
LUN_COLS  = period_cols("Lunas")

def to_jsonb(row, cols):
    return {c.split("_", 1)[1]: float(row[c]) if pd.notna(row[c]) else 0.0 for c in cols if c in df.columns}

def flt(v, default=0.0):
    try:
        f = float(v)
        return f if not pd.isna(f) else default
    except Exception:
        return default

def txt(v):
    if v is None or (isinstance(v, float) and pd.isna(v)):
        return None
    return str(v).strip() or None

# ── Build rows ─────────────────────────────────────────────────────────────────
rows = []
for _, r in df.iterrows():
    est_map  = to_jsonb(r, EST_COLS)
    bm_map   = to_jsonb(r, BM_COLS)
    lun_map  = to_jsonb(r, LUN_COLS)
    rows.append((
        txt(r["KD_CUST"]),          # kd_cust
        txt(r["NM_CUST"]),          # nm_cust
        txt(r["C_URUT"]),           # c_urut
        txt(r.get("KD_SPC")),       # kd_spc
        txt(r.get("NM_SPC")),       # nm_spc
        txt(r.get("ROLE")),         # role
        int(flt(r.get("N_DIVISI", 0))) if r.get("N_DIVISI") else None,  # n_divisi
        txt(r.get("DIV_KODE")),     # div_kode
        flt(r.get("Biaya", 0)),     # biaya
        txt(r.get("PRD_AWAL")),     # prd_awal
        txt(r.get("PRD_AKHIR")),    # prd_akhir
        txt(r.get("NIP_USUL")),     # nip_usul
        txt(r.get("NM_USUL")),      # nm_usul
        txt(r.get("KD_OUTLET")),    # kd_outlet
        txt(r.get("NM_OUTLET")),    # nm_outlet
        txt(r.get("DIV_PROD")),     # div_prod
        txt(r.get("KD_PRODUK")),    # kd_produk
        txt(r.get("NM_PRODUK")),    # nm_produk
        flt(r.get("EstBaris")),     # est_baris
        flt(r.get("BM_BARIS")),     # bm_baris
        sum(est_map.values()),      # total_est
        sum(bm_map.values()),       # total_bm
        sum(lun_map.values()),      # total_lunas
        json.dumps(est_map),        # est_by_period
        json.dumps(bm_map),         # bm_by_period
        json.dumps(lun_map),        # lunas_by_period
        SNAPSHOT_DATE,              # snapshot_date
    ))

# ── Upsert ─────────────────────────────────────────────────────────────────────
UPSERT = """
INSERT INTO "PsspKontrak" (
  id, "kdCust", "nmCust", "cUrut", "kdSpc", "nmSpc", role, "nDivisi", "divKode",
  biaya, "prdAwal", "prdAkhir", "nipUsul", "nmUsul", "kdOutlet", "nmOutlet",
  "divProd", "kdProduk", "nmProduk",
  "estBaris", "bmBaris", "totalEst", "totalBm", "totalLunas",
  "estByPeriod", "bmByPeriod", "lunasByPeriod",
  "snapshotDate", "importedAt"
) VALUES (
  gen_random_uuid(), %s, %s, %s, %s, %s, %s, %s, %s,
  %s, %s, %s, %s, %s, %s, %s,
  %s, %s, %s,
  %s, %s, %s, %s, %s,
  %s::jsonb, %s::jsonb, %s::jsonb,
  %s, NOW()
)
ON CONFLICT ("cUrut", "kdProduk") DO UPDATE SET
  "kdCust"        = EXCLUDED."kdCust",
  "nmCust"        = EXCLUDED."nmCust",
  "kdSpc"         = EXCLUDED."kdSpc",
  "nmSpc"         = EXCLUDED."nmSpc",
  role            = EXCLUDED.role,
  "nDivisi"       = EXCLUDED."nDivisi",
  "divKode"       = EXCLUDED."divKode",
  biaya           = EXCLUDED.biaya,
  "prdAwal"       = EXCLUDED."prdAwal",
  "prdAkhir"      = EXCLUDED."prdAkhir",
  "nipUsul"       = EXCLUDED."nipUsul",
  "nmUsul"        = EXCLUDED."nmUsul",
  "kdOutlet"      = EXCLUDED."kdOutlet",
  "nmOutlet"      = EXCLUDED."nmOutlet",
  "divProd"       = EXCLUDED."divProd",
  "nmProduk"      = EXCLUDED."nmProduk",
  "estBaris"      = EXCLUDED."estBaris",
  "bmBaris"       = EXCLUDED."bmBaris",
  "totalEst"      = EXCLUDED."totalEst",
  "totalBm"       = EXCLUDED."totalBm",
  "totalLunas"    = EXCLUDED."totalLunas",
  "estByPeriod"   = EXCLUDED."estByPeriod",
  "bmByPeriod"    = EXCLUDED."bmByPeriod",
  "lunasByPeriod" = EXCLUDED."lunasByPeriod",
  "snapshotDate"  = EXCLUDED."snapshotDate",
  "importedAt"    = NOW()
"""

BATCH = 500
conn = psycopg2.connect(host=_host, port=_port, dbname=_db, user=_user, password=_pass)
conn.autocommit = False
cur = conn.cursor()

total = len(rows)
done  = 0
for i in range(0, total, BATCH):
    batch = rows[i:i+BATCH]
    psycopg2.extras.execute_batch(cur, UPSERT, batch, page_size=BATCH)
    done += len(batch)
    print(f"  {done}/{total} rows...", end="\r")

conn.commit()
cur.close()
conn.close()
print(f"\nDone. {total} rows upserted into PsspKontrak.")

File: scripts/importPsspKontrak.ts
/**
 * Import PSSP contract history into PsspKontrak table.
 *
 * Source (2026-07-22): "PSSP VB Horizontal dan Vertikal.xlsx", sheet
 * "PSSPVBHor" — replaces the old "20260630_PSSP Pelunasan..." file/format.
 * "PSSPVBVer" (the other sheet in this workbook) is the same data unpivoted
 * to one-row-per-period — redundant for our purposes, not imported.
 *
 * Column differences from the old format:
 *   - No "BM_" (Biaya Murni) columns at all anymore, and no "EstBaris"/
 *     "BM_BARIS" row-total columns — only per-period "Estimasi_YYYYMM" and
 *     "Pelunasan_YYYYMM". estBaris/totalEst are now computed as the sum of
 *     Estimasi_YYYYMM across all period columns (the closest available
 *     substitute for the old dedicated total column). bmBaris/totalBm are
 *     left null and bmByPeriod "{}" — nothing in the app reads them (only
 *     biaya/estBaris/totalLunas are ever queried, see getPsspHistory /
 *     getActivePsspByOutlets in app/actions/customer.ts), so this is safe.
 *   - New "QtyEst_YYYYMM" quantity columns exist but aren't imported — no
 *     schema field or UI consumes a quantity dimension from this table yet.
 *   - "ROLE" column is gone, replaced by "GOL" (golongan/tier, e.g. "USER") —
 *     stored in the `role` field as the closest analog; nothing reads it downstream.
 *   - No date is embedded in this file's name (unlike the old
 *     "20260630_..." naming) — snapshotDate is just the import run's date.
 *
 * This file is large (~30k rows, ~37MB) — run with extra heap:
 *   node --max-old-space-size=6144 node_modules/.bin/tsx scripts/importPsspKontrak.ts [path-to-excel]
 * Default: internal/PSSP VB Horizontal dan Vertikal.xlsx
 *
 * Full-replace semantics (2026-07-22, "ikuti yang baru, yang lama buang" —
 * same principle as the org-struktur/outlet-assignment imports this session):
 * after upserting every row from this run, any PREVIOUSLY-imported PsspKontrak
 * row whose (cUrut, kdProduk) key is NOT in this run's data is deleted — a
 * contract that no longer appears in the latest snapshot is closed/superseded,
 * not still-active stale data.
 */

import "dotenv/config";
import path from "path";
import ExcelJS from "exceljs";
import { Prisma } from "@prisma/client";
import { prisma } from "../src/lib/prisma";

const SHEET_NAME = "PSSPVBHor";

function cellRaw(v: unknown): unknown {
  if (v != null && typeof v === "object" && "result" in (v as object)) {
    return (v as { result: unknown }).result;
  }
  return v;
}

function txt(v: unknown): string | null {
  const raw = cellRaw(v);
  if (raw == null) return null;
  const s = String(raw).trim();
  return s || null;
}

function flt(v: unknown, fallback = 0): number {
  const raw = cellRaw(v);
  if (typeof raw === "number") return isNaN(raw) ? fallback : raw;
  const n = parseFloat(String(raw ?? ""));
  return isNaN(n) ? fallback : n;
}

async function main() {
  const filePath = path.resolve(
    process.argv[2] ?? "internal/PSSP VB Horizontal dan Vertikal.xlsx"
  );

  const snapshotDate = new Date();

  console.log(`Reading: ${filePath} (sheet ${SHEET_NAME}, snapshot ${snapshotDate.toISOString().slice(0, 10)})\n`);

  const wb = new ExcelJS.Workbook();
  await wb.xlsx.readFile(filePath);
  const ws = wb.getWorksheet(SHEET_NAME);
  if (!ws) { console.error(`Sheet "${SHEET_NAME}" not found`); process.exit(1); }

  const headerRow = ws.getRow(1);
  const colIdx = new Map<string, number>();
  for (let c = 1; c <= ws.columnCount; c++) {
    const h = txt(headerRow.getCell(c).value);
    if (h) colIdx.set(h, c);
  }
  const col = (name: string) => colIdx.get(name);

  // Period columns — exact prefix match only (Estimasi_202601, not QtyEst_202601).
  const estCols: { col: number; period: string }[] = [];
  const lunasCols: { col: number; period: string }[] = [];
  for (const [name, c] of colIdx) {
    const estM = name.match(/^Estimasi_(\d{6})$/);
    if (estM) estCols.push({ col: c, period: estM[1] });
    const lunM = name.match(/^Pelunasan_(\d{6})$/);
    if (lunM) lunasCols.push({ col: c, period: lunM[1] });
  }
  console.log(`Period columns — Estimasi: ${estCols.length}, Pelunasan: ${lunasCols.length}\n`);

  type Row = [
    string, string | null, string, string | null, string | null, string | null, number | null, string | null,
    number, string, string, string | null, string | null, string | null, string | null,
    string | null, string, string | null,
    number, number, number,
    string, string,
    Date,
  ];

  // Keyed by cUrut|kdProduk — de-dupes rows sharing the unique constraint so a
  // single multi-row INSERT never targets the same conflict target twice
  // ("last row in the sheet wins", same as a sequential upsert).
  const rowsByKey = new Map<string, Row>();
  let skipped = 0;
  let sourceRowCount = 0;

  for (let rn = 2; rn <= ws.rowCount; rn++) {
    const row = ws.getRow(rn);
    const get = (name: string) => (col(name) ? row.getCell(col(name)!).value : undefined);

    const cUrut = txt(get("C_URUT"));
    const kdCust = txt(get("KD_CUST"));
    if (!cUrut || !kdCust) { skipped++; continue; }
    sourceRowCount++;

    const periodMap = (cols: { col: number; period: string }[]) => {
      const m: Record<string, number> = {};
      for (const { col: c, period } of cols) m[period] = flt(row.getCell(c).value, 0);
      return m;
    };
    const estMap = periodMap(estCols);
    const lunMap = periodMap(lunasCols);
    const estBaris = Object.values(estMap).reduce((a, b) => a + b, 0);
    const totalLunas = Object.values(lunMap).reduce((a, b) => a + b, 0);

    const nDivisiRaw = get("N_DIVISI");
    const nDivisi = nDivisiRaw != null && txt(nDivisiRaw) != null ? Math.trunc(flt(nDivisiRaw)) : null;

    const kdProduk = txt(get("KD_PRODUK")) ?? "";
    rowsByKey.set(`${cUrut}|${kdProduk}`, [
      kdCust, txt(get("NM_CUST")), cUrut, txt(get("KD_SPC")), txt(get("NM_SPC")), txt(get("GOL")), nDivisi, txt(get("DIV_KODE")),
      flt(get("Biaya")), txt(get("PRD_AWAL")) ?? "", txt(get("PRD_AKHIR")) ?? "", txt(get("NIP_USUL")), txt(get("NM_USUL")),
      txt(get("KD_OUTLET")), txt(get("NM_OUTLET")),
      txt(get("DIV_PROD")), kdProduk, txt(get("NM_PRODUK")),
      estBaris, estBaris, totalLunas, // estBaris + totalEst both = sum of Estimasi_YYYYMM (no separate baris total column in this source)
      JSON.stringify(estMap), JSON.stringify(lunMap),
      snapshotDate,
    ]);
  }

  const rows = [...rowsByKey.values()];
  const dupeCount = sourceRowCount - rows.length;
  console.log(`Rows to upsert: ${rows.length} (skipped ${skipped} with missing C_URUT/KD_CUST, ${dupeCount} duplicate cUrut+kdProduk rows collapsed to last value)\n`);

  const BATCH = 500;
  let done = 0;
  for (let i = 0; i < rows.length; i += BATCH) {
    const batch = rows.slice(i, i + BATCH);
    const valueTuples = batch.map((r) => Prisma.sql`(
      gen_random_uuid(), ${r[0]}, ${r[1]}, ${r[2]}, ${r[3]}, ${r[4]}, ${r[5]}, ${r[6]}, ${r[7]},
      ${r[8]}, ${r[9]}, ${r[10]}, ${r[11]}, ${r[12]}, ${r[13]}, ${r[14]},
      ${r[15]}, ${r[16]}, ${r[17]},
      ${r[18]}, NULL, ${r[19]}, NULL, ${r[20]},
      ${r[21]}::jsonb, '{}'::jsonb, ${r[22]}::jsonb,
      ${r[23]}, NOW()
    )`);

    await prisma.$executeRaw`
      INSERT INTO "PsspKontrak" (
        id, "kdCust", "nmCust", "cUrut", "kdSpc", "nmSpc", role, "nDivisi", "divKode",
        biaya, "prdAwal", "prdAkhir", "nipUsul", "nmUsul", "kdOutlet", "nmOutlet",
        "divProd", "kdProduk", "nmProduk",
        "estBaris", "bmBaris", "totalEst", "totalBm", "totalLunas",
        "estByPeriod", "bmByPeriod", "lunasByPeriod",
        "snapshotDate", "importedAt"
      ) VALUES ${Prisma.join(valueTuples)}
      ON CONFLICT ("cUrut", "kdProduk") DO UPDATE SET
        "kdCust" = EXCLUDED."kdCust", "nmCust" = EXCLUDED."nmCust",
        "kdSpc" = EXCLUDED."kdSpc", "nmSpc" = EXCLUDED."nmSpc", role = EXCLUDED.role,
        "nDivisi" = EXCLUDED."nDivisi", "divKode" = EXCLUDED."divKode", biaya = EXCLUDED.biaya,
        "prdAwal" = EXCLUDED."prdAwal", "prdAkhir" = EXCLUDED."prdAkhir",
        "nipUsul" = EXCLUDED."nipUsul", "nmUsul" = EXCLUDED."nmUsul",
        "kdOutlet" = EXCLUDED."kdOutlet", "nmOutlet" = EXCLUDED."nmOutlet",
        "divProd" = EXCLUDED."divProd", "nmProduk" = EXCLUDED."nmProduk",
        "estBaris" = EXCLUDED."estBaris", "bmBaris" = EXCLUDED."bmBaris",
        "totalEst" = EXCLUDED."totalEst", "totalBm" = EXCLUDED."totalBm", "totalLunas" = EXCLUDED."totalLunas",
        "estByPeriod" = EXCLUDED."estByPeriod", "bmByPeriod" = EXCLUDED."bmByPeriod", "lunasByPeriod" = EXCLUDED."lunasByPeriod",
        "snapshotDate" = EXCLUDED."snapshotDate", "importedAt" = NOW()
    `;

    done += batch.length;
    console.log(`  ${done}/${rows.length} rows...`);
  }

  console.log(`\n✅ Upserted ${rows.length} rows into PsspKontrak.`);

  // ── Full-replace: retire contracts no longer present in this snapshot ──────
  console.log("\nChecking for stale contracts to retire...");
  const newKeys = new Set(rows.map((r) => `${r[2]}|${r[16]}`)); // cUrut|kdProduk
  const existing = await prisma.psspKontrak.findMany({ select: { id: true, cUrut: true, kdProduk: true } });
  const staleIds = existing
    .filter((e: { id: string; cUrut: string; kdProduk: string | null }) => !newKeys.has(`${e.cUrut}|${e.kdProduk ?? ""}`))
    .map((e: { id: string }) => e.id);

  if (staleIds.length > 0) {
    console.log(`  Deleting ${staleIds.length} stale rows (contract+product no longer in this snapshot)...`);
    for (let i = 0; i < staleIds.length; i += BATCH) {
      const chunk = staleIds.slice(i, i + BATCH);
      await prisma.psspKontrak.deleteMany({ where: { id: { in: chunk } } });
      console.log(`    ${Math.min(i + BATCH, staleIds.length)}/${staleIds.length}`);
    }
  } else {
    console.log("  None — nothing to retire.");
  }

  console.log(`\n✅ Done. ${rows.length} rows current, ${staleIds.length} stale rows retired.`);
  await prisma.$disconnect();
}

main().catch((e) => {
  console.error(e);
  prisma.$disconnect();
  process.exit(1);
});
